sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,1059 @@
+        return validateClassesEquality(
+            inputClass,
+            conf.getHadoopConfiguration().getClass(HadoopInputFormatIOConstants.KEY_CLASS,
+                Object.class));
+      } catch (ClassNotFoundException e) {
+        /*
+         * Given inputFormatGenericClassName is a type parameter i.e. T, K, V, etc. In such cases
+         * class validation for user provided input key will not work correctly. Therefore the need
+         * to validate key class by encoding and decoding key object with the given coder.
+         */
+        return validateClassUsingCoder(keyObject, coder);
+      }
+    }
+
+    /**
+     * Returns true if value class set by the user is compatible with the value class of a pair
+     * returned by RecordReader. User provided value class is validated against the parameterized
+     * type's type arguments of InputFormat. If parameterized type has any type arguments such as T,
+     * K, V, etc then validation is done by encoding and decoding value object of first pair
+     * returned by RecordReader.
+     */
+    private <T> boolean validateValueClass(String inputFormatGenericClassName, Coder coder,
+        Object valueObject) {
+      try {
+        Class<?> inputClass = Class.forName(inputFormatGenericClassName);
+        /*
+         * Validates value class with InputFormat's parameterized type.
+         */
+        return validateClassesEquality(
+            inputClass,
+            conf.getHadoopConfiguration().getClass(HadoopInputFormatIOConstants.VALUE_CLASS,
+                Object.class));
+
+      } catch (Exception e) {
+        /*
+         * Given inputFormatGenericClassName is a type parameter i.e. T, K, V, etc. In such cases
+         * class validation for user provided input value will not work correctly. Therefore the
+         * need to validate value class by encoding and decoding value object with the given coder.
+         */
+        return validateClassUsingCoder(valueObject, coder);
+      }
+    }
+
+    /**
+     * Validates if key/value object encodes and decodes successfully using the given coder.
+     */
+    private <T> boolean validateClassUsingCoder(Object entity, Coder<T> coder) {
+      return checkEncodingAndDecoding(coder, (T) entity);
+    }
+
+    /**
+     * Validates if a given class is type of a given expectedClass.
+     */
+    private boolean validateClassesEquality(Class<?> expectedClass, Class<?> actualClass) {
+      return actualClass.isAssignableFrom(expectedClass);
+    }
+
+    /**
+     * Validates whether the input gets encoded or decoded correctly using the provided coder.
+     */
+    private <T> boolean checkEncodingAndDecoding(Coder<T> coder, T input) {
+      try {
+        CoderUtils.clone(coder, input);
+      } catch (CoderException e) {
+        return false;
+      }
+      return true;
+    }
+
+    /**
+     * Returns parameterized type of the InputFormat class.
+     */
+    private ParameterizedType determineGenericType() {
```

Comment:
It is considered that user provided inputFormatClass will always inherit from ParameterizedType. Have added the comment accordingly.